plutus-core/plutus-ir/Language/PlutusIR/TypeCheck/Internal.hs

```diff
@@ -314,18 +317,70 @@ checkTypeFromBinding recurs = \case
                       -- OPTIMIZE: we use inferKind for scope-checking, but a simple ADT-traversal would suffice
                       for_ (funTyArgs ty) inferKindM
 
+-- | Check that the in-Term's inferred type of a Let has kind *.
+-- Skip this check at the top-level, to allow top-level types to escape; see Note [PIR vs paper FIR Difference-2].
+checkStarInferred :: AsTypeError e term uni ann
+                  => ann -> Normalized (Type TyName uni b) -> TypeCheckM uni IsTop e ()
+checkStarInferred ann t = do
+    isTop <- asks view tceOther
+    case isTop of
+        NoTop  -> checkKindM ann (ann <$ unNormalized t) $ Type ()
+        -- NOTE: we completely skip the check in case of toplevel because we would need an *final, extended Gamma environment*
```

Comment:
It took me a few tries to get what you're saying here. I think what you're saying is: even in the case where we let the type variables escape, the kind of the type should still be star - it's just that in order to check that we'd have to use the variables which are no longer in scope.

This is an interesting point! It suggests an alternative approach: we could keep the check in both cases (as you suggest we do), but simply change *where* we put it. For example, we could do
```
    foldr checkBindingThenScope (inferTypeM inTerm >>= through (checkStarInferred ann)) bs
```

Having written that down, though, I see why it's wrong. If we have something like this:
```
let x = ... in let datatype Bool = ... in True
```
then the version I suggested would fail this at the outermost let binding.

So it's not enough to have the type environment with all the bindings in the current let binding, you'd also have to have them for *inner* let bindings too. So I don't think we can do anything better here.

Overall: worth a note!

Reply:
Yes, exactly. I copy your words to a note.

Reply:
Ah, ok, so my naive 

> Like, can we check the kind of the type returned by `inferTypeM inTerm`` right there in the accumulator? And if escaped type are allowed, then simply set the context to the one we had before we started doing `checkBindingThenScope`.

doesn't work. Well:

1. you wouldn't want to check that the inferred type is well-kinded each time you leave a `let` anyway as it's just wasteful to do the same work over and over again
2. you could simply return the context along with the inferred type and check the type once you leave the outer `let`
3. or perhaps instead of having `YesEscape`/`NoEscape` you can store in your `Reader` a continuation that you can use to kind check an inferred type. And also store the context, in which the continuation was created. And once you reach a leaf of the AST, say, a variable, you invoke that continuation on the inferred type of the variable. If you're at the top level, then you kind check the inferred type in the current context, otherwise you set the current context to the stored original one. Or something along these lines (I don't think having a continuation is a necessity, you could probably get away with just storing the original context in `NoEscape`, but with CPS it should be possible to use types to make sure that the additional kind checking happens at every possible kind of a leaf, like variables, constants, error, whatever). I haven't worked out the details, obviously, so take it with a grain of salt

Reply:
Note that the issue isn't that we'll get a type which hasn't been kind-checked and could be nonsensical. That is: I believe that the type of a let-binding is *guaranteed* to be correctly of kind `*`... except possibly if some of the variables which it refers to are now out of scope. So really the *only* function of this check is to prevent type variables from escaping. So if we're omitting it, there's no point trying to do it later or something.

(I don't have a proof of this, but I also can't think of how it could go wrong otherwise.)

Reply:
That makes sense to me. We also don't perform an analogous check in the Plutus Core type checker. I'm convinced then, but let's have a note about that.